[clock] Documenting ways to access time (#8729)
MystenLabs/sui: doc/in-progress/time.md

# Accessing Time in Move

> 

Author: Sounds good!  Let me pull it out into a suggestion so I can apply it.  I'm not attached to the structure of this document by the way -- it's possible that this content might be better off as a sub-section of some other page, or split across a few pages, or what have you, so I'm just staging it in the `in-progress` section, for you and @randall-Mysten to take from there, and find a home in our docs.

```suggestion

You have options when needing to access network-based time for your transactions. If you need a near real-time measurement (within a few seconds), use the immutable reference of time provided by the `Clock` module in Sui Move. The reference value from this module updates with every network checkpoint. If you don't need as current a time slice, use the `epoch_timestamp_ms` function to capture the precise moment the current epoch started.
```

## `sui::clock::Clock`

> Transactions that need access to an accurate clock must pass a read-only reference of `sui::clock::Clock` in as an entry function parameter.  The only available instance is found at address `0x6`.

Author: Added an example, let me know what you think!

A unix timestamp in milliseconds can be extracted from an instance of `Clock` using

```
module sui::clock {
    public fun timestamp_ms(clock: &Clock): u64;
}
```

> **The resulting value is expected to change every 2 to 3 seconds**, at the rate that checkpoints are committed by the network.

Reviewer: ```suggestion
**Expect the resulting value to change every 2 to 3 seconds**, at the rate the network commits checkpoints.
```

> Successive calls to `sui::clock::timestamp_ms` in the same transaction will always produce the same result (transactions are considered to take effect instantly), and may also overlap with other transactions that touch the same shared objects, so it is not safe to assume that time progresses between transactions that are sequenced relative to each other.

Reviewer: ```suggestion
Successive calls to `sui::clock::timestamp_ms` in the same transaction always produce the same result (transactions are considered to take effect instantly), and might also overlap with other transactions that touch the same shared objects. Consequently, you can't assume that time progresses between transactions that are sequenced relative to each other.
```

> Any transaction that requires access to a `Clock` must go through **consensus**, because the only available instance is a **shared object**, so this technique is not suitable for transactions that must use the single-owner fast-path (see "Epoch timestamps" for a single-owner-compatible source of timestamps).

Reviewer: ```suggestion
Any transaction that requires access to a `Clock` must go through [consensus](/learn/architecture/consensus) because the only available instance is a [shared object](/learn/objects#shared). As a result, this technique is not suitable for transactions that must use the single-owner fast-path (see [Epoch timestamps](#epoch-timestamps) for a single-owner-compatible source of timestamps).
```

Reviewer: I will update these links to github paths once this page has a home.

> **Transactions that use the clock must accept it as an immutable reference** (not a mutable reference or value).  This prevents contention, as transactions that access the `Clock` can only read it, so do not need to be sequenced relative to each other.  Validators will refuse to sign transactions that do not meet this requirement and packages that include entry functions that accept a `Clock` or `&mut Clock` will fail to publish.

Reviewer: ```suggestion
**Transactions that use the clock must accept it as an immutable reference** (not a mutable reference or value).  This prevents contention, as transactions that access the `Clock` can only read it, so do not need to be sequenced relative to each other.  Validators refuse to sign transactions that do not meet this requirement and packages that include entry functions that accept a `Clock` or `&mut Clock` fail to publish.
```

> Code that depends on `Clock` can be tested using

Reviewer: ```suggestion
The following example shows how to test 'Clock'-dependent code by manually incrementing its timestamp, which is possible only in test code: 
```

```
module sui::clock {
    #[test_only]
    public fun increment_for_testing(clock: &mut Clock, tick: u64);
}
```

> which can increment the timestamp in the `Clock` in test code.

Reviewer: ```suggestion

```

## Epoch timestamps

> All transactions (including ones that do not go through consensus) can access the timestamp for the start of the current epoch, via the following function:

Reviewer: ```suggestion
You can use the following function to access the timestamp for the start of the current epoch for all transactions (including ones that do not go through consensus):
```

```
module sui::tx_context {
    public fun epoch_timestamp_ms(ctx: &TxContext): u64;
}
```

> Which returns the point in time when the current epoch started, as a millisecond granularity unix timestamp in a `u64`.  **This value changes roughly once every 24 hours**, when the Epoch changes.

Reviewer: ```suggestion
The preceding example returns the point in time when the current epoch started, as a millisecond granularity unix timestamp in a `u64`.  **This value changes roughly once every 24 hours**, when the epoch changes.
```

> The value returned by this function can be updated in tests for time-sensitive code that use `sui::test_scenario`, using

Reviewer: ```suggestion
You can use the value the previous function returns to update tests for time-sensitive code that uses `sui::test_scenario`. For example,
```

Author: This one is the other way around -- i.e. when testing code that uses the previous function, we can use the function below to update the value it returns.

```suggestion
Tests based on `sui::test_scenario` for time-sensitive code that uses the previous function can use `sui::test_scenario::later_epoch` to simulate the progress of time:
```

Author: @ronny-mysten, I made an attempt to reword this while preserving meaning, but it does sound pretty clunky, flagging for you in case there's a better way to say it.

Reviewer 2: You can use the `sui::test_scenario::later_epoch` function to simulate time progression to test time-sensitive code based on `sui::test_scenario`.

Reviewer 2: yeah this is a tough one

```
module sui::test_scenario {
    public fun later_epoch(
        scenario: &mut Scenario,
        delta_ms: u64,
        sender: address,
    ): TransactionEffects;
}
```

> which behaves like `sui::test_scenario::next_epoch` (finishes the current transaction and epoch), but also increments the timestamp by `delta_ms` milliseconds.

Reviewer: ```suggestion
The preceding code behaves like `sui::test_scenario::next_epoch` (finishes the current transaction and epoch), but also increments the timestamp by `delta_ms` milliseconds.
```

**Transactions that use the clock must accept it as an immutable reference** (not a mutable reference or value).  This prevents contention, as transactions that access the `Clock` can only read it, so do not need to be sequenced relative to each other.  Validators refuse to sign transactions that do not meet this requirement and packages that include entry functions that accept a `Clock` or `&mut Clock` fail to publish.

The following function is used to test 'Clock'-dependent code by manually incrementing its timestamp, which is possible only in test code: 

```
module sui::clock {
    #[test_only]
    public fun increment_for_testing(clock: &mut Clock, tick: u64);
}
```

## Epoch timestamps

You can use the following function to access the timestamp for the start of the current epoch for all transactions (including ones that do not go through consensus):

```
module sui::tx_context {
    public fun epoch_timestamp_ms(ctx: &TxContext): u64;
}
```

The preceding function returns the point in time when the current epoch started, as a millisecond granularity unix timestamp in a `u64`.  **This value changes roughly once every 24 hours**, when the epoch changes.

> Tests based on `sui::test_scenario` can use `later_epoch` (below), to exercise time-sensitive code that uses `epoch_timestamp_ms` (above):

Reviewer: ```suggestion
Tests based on `sui::test_scenario` can use `later_epoch` (following code), to exercise time-sensitive code that uses `epoch_timestamp_ms` (previous code):
```
Nit. Can't assume these will always be physically above or below this sentence.
